libbeat/processors/add_id/generator/es_generator.go

```diff
@@ -0,0 +1,121 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type esTimeBasedUUIDGenerator struct{}
+
+// Singleton instance
+var _esTimeBasedUUIDGenerator IDGenerator = (*esTimeBasedUUIDGenerator)(nil)
+
+// ESTimeBasedUUIDGenerator returns the singleton instance for this generator
+func ESTimeBasedUUIDGenerator() IDGenerator {
+	return _esTimeBasedUUIDGenerator
+}
+
+var (
+	sequenceNumber uint32
+	lastTimestamp  time.Time
+	mac            []byte
+	mu             sync.Mutex
+)
+
+func init() {
+	m, err := getSecureMungedMACAddress()
+	if err != nil {
+		panic(err)
+	}
+	mac = m
+	sequenceNumber = rand.Uint32()
+}
```

Comment:
I think I commented on this before, but the comment seems to be gone.

By default the go runtime does not initialize the `rand` package. Which means that each beat runs with the very same seed and each Beat is likely to return the same random number here. Unless some other package has intialized the rand package by chance.

The getSecureMungedMACAdress might also run into issues during startup. Better to defer the initialization until you need to generate the first ID. Protecting the initialization using a `sync.Once` object should do the trick.

Reply:
Yes, a few older comments got outdated due to the rename 😕. They are still visible in the PR though (it's just not possible to find them in the file diff view) so I haven't forgotten about them.